File: day02/day02b.cpp
#include <cstdio>
#include <fstream>
#include <iostream>
#include <string>

bool validate_line(std::string &line) {
  // Parse line into min and max occurrences, desired char, and password
  int indexL = 0;
  int indexR = 0;
  char desired_char = 0;
  char *password = new char[line.length() + 1];

  if(std::sscanf(line.c_str(), "%d-%d %c: %s", &indexL, &indexR, &desired_char, password) != 4) {
    return false;
  }

  // Switch from 1-indexing to 0-indexing
  indexL--;
  indexR--;

  int char_amt = (password[indexL] == desired_char) + (password[indexR] == desired_char);
  return (char_amt == 1);
}

int main() {
  // Read input from file
  std::ifstream file("day02.txt");
  std::string line;

  // Count whether each line is valid
  int count = 0;
  while(std::getline(file, line)) {
    if(validate_line(line))
      count++;
  }
  
  std::cout << count << std::endl;

  return 0;
}

File: day02/day02b.py
#!/usr/bin/python3

import parse

def validate_line(line):
    # Parse line into left and right index, desired char, and password
    parsed = parse.parse("{}-{} {}: {}", line)
    indexL = int(parsed[0]) - 1 # Switch from 1-indexing to 0-indexing
    indexR = int(parsed[1]) - 1
    char = parsed[2]
    password = parsed[3]

    char_amt = (password[indexL] == char) + (password[indexR] == char)
    return char_amt == 1

with open('day02.txt', 'r') as f:
    num_valid = sum(1 for line in f if validate_line(line))
    print(num_valid)
